Allow scanning multiple ports on build nodes

nre-ableton opened the pull request:
This PR changes the way that the `jenkins-node-scanner.py` script works with respect to the `--target-port` argument. Specifically, this argument may now be specified multiple times in order to scan multiple ports, and the ports are tested with a basic TCP socket connection to ensure that they are responding. Only ports which are actually active will be added to the generated output file. Also, the `--target-port` argument *must* now be specified at least once, and the default value (previously `9100`) has been removed.

Ping @AbletonDevTools/gotham-city

nre-ableton committed:
Exit the script if the period is 0

This is useful for debugging purposes.

nre-ableton committed:
Fix groovylint brace violation

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -231,7 +232,10 @@ def main():
+
+        if args.period:
+            time.sleep(args.period)
+        break
```
This `break` is done unconditionally. There is no loop anymore.

nre-ableton replied:
Right. 😅 Well, this PR is still technically WIP...

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -231,7 +232,11 @@ def main():
                 write_output(args.output_file, node_info)
 
         logging.debug('Waiting %d seconds', args.period)
```
Please move the logging statement in the `if args.period:` block.

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -221,13 +221,18 @@ def main():
+                node_info = []
+
+                for node, ip_addr in node_ips:
+                    targets: ['%s:%d' % (ip_addr, args.target_port)],
```
Did you mean
```python
                    targets = ['%s:%d' % (ip_addr, args.target_port)]
```
?

nre-ableton replied:
I'll inline this instead

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -221,13 +221,18 @@ def main():
+                        'labels': {
+                            'jenkins_master': args.url,
+                            'node': node['name'],
+                        },
```
```python
                        'labels': {'jenkins_master': args.url, 'node': node['name']},
```
fits in 90 columns.

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -224,7 +227,10 @@ def main():
+                    targets = []
+
+                    for port in args.target_port:
+                        targets.append('%s:%d' % (ip_addr, port))
```
This loop can be expressed as a list comprehension:
```python
                    targets = ['%s:%d' % (ip_addr, port) for port in args.target_port]
```

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -230,7 +242,8 @@ def main():
                     for port in args.target_port:
-                        targets.append('%s:%d' % (ip_addr, port))
+                        if test_port(ip_addr, port):
+                            targets.append('%s:%d' % (ip_addr, port))
```
This loop can still be expressed as a list comprehension:
```python
                    targets = ['%s:%d' % (ip_addr, port) for port in args.target_port
                               if test_port(ip_addr, port)]
```

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -165,6 +170,17 @@ def get_nodes(master):
         return []
 
 
+def test_port(ip_addr, port):
```
IMHO, `is_port_open` would be a better name, since this function returns a boolean.

ala-ableton reviewed (commented):
@nre-ableton the final diff is looking way better now!
Please rebase, squash and reword (especially 881a267c056648dd997dd7fcffec64910a759eb1) and I'll do another review round afterwards, thanks!

mst-ableton commented on jenkins_node_scanner.py:
```diff
@@ -220,18 +236,23 @@ def main():
+                    node_info.append({
+                        'labels': {'jenkins_master': args.url, 'node': node['name']},
+                        'targets': ['%s:%d' % (ip_addr, port) for port in args.target_port
+                                    if test_port(ip_addr, port)],
```
If neither port is available, it will just emit an empty `targets` field. I guess that's okay.

mst-ableton reviewed (commented):
flattened code looks fine, will wait until rebase + cleanup before doing final review

nre-ableton committed:
Use shutil.move instead of os.rename

os.rename will raise an error if the files are on different devices,
which is often the case on Linux systems which mount /tmp in memory.

nre-ableton committed:
Let --target-port option be used multiple times

This allows for scanning multiple prometheus exporters on build nodes.

nre-ableton committed:
Test TCP ports before adding them to the list

nre-ableton commented:
Ok, this PR is now squashed and rebased. However, I’ve made some additional fixes after doing further testing of this branch.

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -100,9 +100,12 @@ def get_args():
-        default=9100,
+        action='append',
+        default=[],
+        required=True,
```
Passing `metavar='target_ports',`  here would allow to write `for port in args.target_ports` on line 233.

nre-ableton replied:
I don't really understand this comment. I'm not familiar with `metavar`, so I looked up [the documentation](https://docs.python.org/dev/library/argparse.html#metavar) and found this:

> Note that metavar only changes the displayed name - the name of the attribute on the parse_args() object is still determined by the dest value.

I don't think this does what you think it does...

Also, line 233 is a blank line. Did you mean line 245? So in other words:

```python
node_info = [{
    'labels': {
        'jenkins_master': args.url,
        'node': node['name'],
    },
    'targets': ['%s:%d' % (ip_addr, port) for port in args.target_ports  # <-- here!
                if is_port_open(ip_addr, port)],
} for node, ip_addr in node_ips]
```

If that's what you meant, then doesn't work. It throws `AttributeError: 'Namespace' object has no attribute 'target_ports'`.

_Edit_: I think you must have meant the `dest` parameter instead of `metavar`. Fixed in 9bba8de.

ala-ableton replied:
Oh yeah, I meant `dest`. Sorry for the confusion.

ala-ableton commented on jenkins_node_scanner.py:
```diff
@@ -165,6 +171,17 @@ def get_nodes(master):
 
 
+def is_port_open(ip_addr, port):
+    """Test that a TCP port is open on the given IP address."""
```
```suggestion
    """Return whether a TCP port is open on the given IP address."""
```

ala-ableton reviewed (changes requested):
Typo in 5030b985e7d1eec0741a3d83598f2dbd8198e844's commit message: `Linix` -> `Linux`.

nre-ableton commented:
@ala-ableton @mst-ableton Are you both ok with me rebasing/squashing this PR again? If so, I'll do so and also take the opportunity to fix the typo in 5030b985e7d1e.

ala-ableton reviewed (commented):
@nre-ableton feel free to rebase

nre-ableton commented:
@ala-ableton Done! Also, I added a fix to the `Jenkinsfile`, which I just realized I would have broken given that `--target-port` is now a required argument. 😅

ala-ableton commented on Jenkinsfile:
```diff
@@ -48,6 +48,8 @@ devToolsProject.run(
       '8000',
       '-v jenkins-nodes:/jenkins_nodes',
+      '--target-port 9100',
+      '--target-port 9323',
```
`deploy` takes a maximum of 3 arguments, here we are passing 5.

nre-ableton replied:
Fixed in de1ea21.

mst-ableton commented on jenkins_node_scanner.py:
```diff
@@ -231,8 +232,11 @@ def main():
 
-        logging.debug('Waiting %d seconds', args.period)
-        time.sleep(args.period)
+        if args.period:
```
I'd actually appreciate if this logic was inverted:

```
if args.period == 0:
    break


logging.debug('Waiting %d seconds', args.period)
time.sleep(args.period)

nre-ableton replied:
Hmm, I kinda like this compared to the `if`/`else`. Fixed in 000aa6d78762dc4c670017480ff14c3895ff7f14.

mst-ableton reviewed (approved):
lgtm, I have a small change but it's okay without it too

ala-ableton reviewed (approved):
Once rebased, LGTM :+1:

nre-ableton commented:
@ablbot autosquash
